pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/MaximumPinotConsumptionDelayTracker.java

```diff
@@ -0,0 +1,344 @@
+  private static final int MAX_NUMBER_OF_TIME_TICKS_BEFORE_VERIFICATION = 5;
+  // Sleep interval for timer thread that triggers read of ideal state
+  private static final int TIMER_THREAD_TICK_INTERVAL_MS = 60000;
+  // Delay Timer thread for this time after starting timer
+  private static final int INITIAL_TIMER_THREAD_DELAY_MS = 1000;
+  // Mutable versions of the above constants so we can test with smaller delays
+  int _maxNumberOfTicksBeforeVerification;
+  int _timerThreadTickIntervalMs;
+  int _initialTimeThreadDelayMs;
+
+  // The following two fields are used to track the current maximum delay and the reporting partition.
+  private long _maxPinotConsumptionDelay;
+  private long _maxPartitionId;
+
+  // Timer task to check partitions that are inactive against ideal state.
+  private TrackingTimerTask _timerTask;
+  private Timer _timer;
+
+  private ServerMetrics _serverMetrics;
+  private String _tableNameWithType;
+
+  private boolean _enableAging;
+
+  RealtimeTableDataManager _realTimeTableDataManager;
+
+  /*
+   * Helper function to initialize maximum to values that are replaced by the first sample.
+   */
+  private void initMaxPinotConsumptionDelay() {
+    _maxPinotConsumptionDelay = -1;
+    _maxPartitionId = -1;
+  }
+
+  /*
+   * Helper function to update the maximum when the current maximum is removed or updated.
+   * If no samples left we set maximum to minimum so new samples can be recorded.
+   */
+  private void updateMaximum() {
+    DelayMeasure newMax = null;
+    for (long pid : _partitionToDelaySampleMap.keySet()) {
+      DelayMeasure currentMeasure = _partitionToDelaySampleMap.get(pid);
+      if ((newMax == null) || (currentMeasure._delayMilliseconds > newMax._delayMilliseconds)) {
+        _maxPartitionId = pid;
+        newMax = currentMeasure;
+      }
+    }
+    if (newMax != null) {
+      _maxPinotConsumptionDelay = newMax._delayMilliseconds;
+    } else {
+      // Nothing left for this table so init to minimum
+      initMaxPinotConsumptionDelay();
+    }
+  }
+
+  /*
+   * This function will read ideal state for the given partition and return true if ideal state reflects the
+   * partition must be hosted by the current server, false otherwise.
+   *
+   * @param partitionId Partition ID that we want to verify is still hosted by this server.
+   */
+  private boolean isPartitionHostedByThisServerPerIdealState(long partitionId) {
+    return _realTimeTableDataManager.isPartitionHostedInThisServer(partitionId);
+  }
+
+  /*
+   * Helper function to be called when we should stop tracking a given partition. Removes the partition from
+   * all our maps, it also updates the maximum if the tracked partition was the previous maximum.
+   *
+   * @param partitionId partition ID which we should stop tracking.
+   */
+  private void removePartitionId(long partitionId) {
+    if (_partitionToDelaySampleMap.containsKey(partitionId)) {
+      _partitionToDelaySampleMap.remove(partitionId);
```

Comment:
Calling `remove()` should return the previous value

Reply:
Yes, I will remove all contains as it does not hurt to remove anyways